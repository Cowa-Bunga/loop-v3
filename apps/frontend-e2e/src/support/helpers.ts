export const testIdSelector = (testId: string) => `[data-testid="${testId}"]`

export const COOKIE_NAME = 'cowabunga-user-cookie'
