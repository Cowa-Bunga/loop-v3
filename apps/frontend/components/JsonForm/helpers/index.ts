const makeInputId = (id: string, suffix?: string) => {
  const parts = id.split('/')

  if (suffix) {
    return `${parts[parts.length - 1]}-${suffix}`
  }

  return parts[parts.length - 1]
}

export { makeInputId }
