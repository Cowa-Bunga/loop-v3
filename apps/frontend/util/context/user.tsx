import { createContext, useContext, useEffect } from 'react'
import { useMergeState } from '@hooks'
import { IUserContext } from '@util/types/IappContext'
import { useSession } from 'next-auth/react'
import { authFirebase } from '@util/lib/firebase'
import { getAuth } from 'firebase/auth'
import { useFirebaseApp } from 'reactfire'

const UserContext = createContext({
  state: {} as IUserContext,
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  update: (updated: object) => {}
})

export function UserWrapper({ children }) {
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const initialState = {} as IUserContext
  const [state, updateState] = useMergeState(initialState)
  const firebaseAuth = getAuth(useFirebaseApp())
  const { data } = useSession()
  const update = (updated: object) => {
    updateState(updated as IUserContext)
  }

  useEffect(() => {
    // temp caching method for user context to persist beyond browser reload...
    // need to find a better way to do this as this data will become stale
    if (JSON.stringify(initialState) !== JSON.stringify(state)) {
      localStorage.setItem('userContext', JSON.stringify(state))
    }
  }, [initialState, state])

  useEffect(() => {
    const userContext = localStorage.getItem('userContext')
    if (userContext) {
      updateState(JSON.parse(userContext))
    }

    if (data) {
      const session = {
        ...data.user
      } as ISessionUser
      if (session.firebase_token) {
        authFirebase(firebaseAuth, session.firebase_token)
      }
    }
  }, [])

  return (
    <UserContext.Provider value={{ state, update }}>
      {children}
    </UserContext.Provider>
  )
}

export function useUserContext() {
  return useContext(UserContext)
}
