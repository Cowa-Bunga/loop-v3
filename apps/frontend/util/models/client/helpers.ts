import { defaultsDeep } from 'lodash'
import { MODEL_CLIENT_EMPTY } from './model'

export const modelClientList = (
  data: { [s: string]: unknown } | ArrayLike<unknown>
): IClient[] => {
  return Object.entries(data)
    .map(modelClientData)
    .map((el: never) => el['1'])
    .sort((a: IClient, b: IClient) => a.name.localeCompare(b.name))
}

export const modelClientData = (data: unknown): IClient => {
  return defaultsDeep(data, MODEL_CLIENT_EMPTY)
}
