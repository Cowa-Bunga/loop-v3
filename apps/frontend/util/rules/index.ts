// business rules engine

export const rules = {}
