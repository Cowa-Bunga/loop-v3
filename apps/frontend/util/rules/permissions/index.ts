export const MODEL_LEGACY_PERMISSIONS = {
  name: 'Permissions',
  primaryKey: 'id',
  properties: {
    id: 'string',
    fleet: 'bool',
    administrator: 'bool',
    scopes: ['orders:create', 'orders:recreate', 'import:create']
  }
}

// WIP - TODO: Add all permissions and share library with backend
export const MODEL_PERMISSIONS = [
  // ----- ADMIN
  'admin:all',
  'admin:clients',
  'admin:users',
  'admin:permissions',
  'admin:settings',
  'admin:logs',
  'admin:reports',
  'admin:billing',
  'admin:notifications',
  'admin:alerts',
  'admin:alerts:all',
  'admin:alerts:clients',
  'admin:alerts:users',
  'admin:alerts:permissions',
  'admin:alerts:reports',
  'admin:alerts:billing',
  'admin:alerts:notifications',

  // ----- DRIVER
  'driver:all',
  'driver:clients',

  // ----- FLEET
  'fleet:all',
  'fleet:clients',

  // ----- MODULES
  'mod:all',
  'mod:map',
  'mod:reports',
  'mod:intelligence',
  'mod:alerts',
  'mod:notifications',
  'mod:billing',

  // ---- INTEGRATIONS
  'int:all',
  'int:google:traffic',
  'int:google:reports',
  'int:google:maps',
  'int:google:geocoding',
  'int:google:places',
  'int:google:directions',
  'int:google:distance',
  'int:google:geolocation',
  'int:google:geofence',

  'int:dataflow:all',
  'int:dataflow:clients',

  // ----- CLIENT
  'client:all',
  'client:map',
  'client:tracking'
]

export const checkPermissions = (permissions: string[], required: string[]) => {
  return required.every((permission) => permissions.includes(permission))
}
