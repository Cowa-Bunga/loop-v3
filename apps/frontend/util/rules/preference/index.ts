import { deepmerge } from '@mui/utils'

// This file is for user preference rules
// there are two data sources for preferences, the api/db and the device
// device related rules may be small things from light/dark mode or further integrated items

// util method
export function hasPreference(rule: string) {
  return !!sdk.get(rule)
}

// management sdk
const sdk = {
  cache: {},

  init: () => {
    // TODO: encrypt/decrypt
    let local = JSON.parse(localStorage.getItem('loop-preferences'))
    local = local || {}
    // TODO: get db/api preferences to join
    const server = {}
    sdk.cache = { ...server, ...local, init: new Date() }
  },

  get: (key?: string) =>
    key ? (sdk.data[key] ? sdk.data[key] : false) : sdk.data,

  set: (data) => {
    sdk.data = deepmerge(sdk.data, data)
    return sdk.data
  },
  data: {}
}

//nb: not test
