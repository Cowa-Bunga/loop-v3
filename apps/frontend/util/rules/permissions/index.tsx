import { createContext } from 'react'
import { useUserContext } from '@context/user'
import { useContext } from '@hooks'

export const MODEL_LEGACY_PERMISSIONS = {
  name: 'Permissions',
  primaryKey: 'id',
  properties: {
    id: 'string',
    fleet: 'bool',
    administrator: 'bool',
    scopes: ['admin:all', 'admin:clients', 'admin:users']
  }
}

export const checkPermissions = (permissions: string[], required: string[]) => {
  return required.every((permission) => permissions.includes(permission))
}

interface IPermissionContext {
  canAccess: (required: string[]) => boolean
}

const PermissionContext = createContext<IPermissionContext>({
  canAccess: (required: string[]) => false
})

const PermissionsProvider = ({ children }) => {
  const {
    state: { permissions }
  } = useUserContext()

  const scopes = new Set(permissions.scopes || [])

  if (permissions.fleet) {
    scopes.add('fleet:access')
  }

  if (permissions.administrator) {
    scopes.add('admin:access')
  }

  const canAccess = (required: string[]) =>
    checkPermissions([...scopes], required)

  return (
    <PermissionContext.Provider value={{ canAccess }}>
      {children}
    </PermissionContext.Provider>
  )
}

function usePermissions() {
  return useContext(PermissionContext)
}

export { usePermissions, PermissionsProvider }
