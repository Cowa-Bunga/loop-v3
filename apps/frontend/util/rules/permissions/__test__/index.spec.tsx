import { PermissionsProvider, usePermissions } from '@util/rules/permissions'
import { renderWithProviders } from '@test/helpers'
import { useUserContext } from '@context/user'

const mockUseUserContext = useUserContext as jest.Mock
jest.mock('@context/user', () => ({
  useUserContext: jest.fn()
}))

const TestingComponent = () => {
  const { canAccess } = usePermissions()
  return (
    <>
      {canAccess(['fleet:access']) && <div>fleet</div>}
      {canAccess(['admin:access']) && <div>admin</div>}
      {canAccess(['order:create']) && <div>order</div>}
    </>
  )
}

describe('<PermissionsProvider />', () => {
  it('should render the fleet component when the user has fleet access', () => {
    mockUseUserContext.mockReturnValue({
      state: {
        permissions: {
          fleet: true
        }
      }
    })
    const { container } = renderWithProviders(
      <PermissionsProvider>
        <TestingComponent />
      </PermissionsProvider>
    )
    expect(container).toMatchSnapshot()
  })

  it('should render the fleet component when the user has admin access', () => {
    mockUseUserContext.mockReturnValue({
      state: {
        permissions: {
          administrator: true
        }
      }
    })
    const { container } = renderWithProviders(
      <PermissionsProvider>
        <TestingComponent />
      </PermissionsProvider>
    )
    expect(container).toMatchSnapshot()
  })

  it('should render the order create component when the order:create has admin access', () => {
    mockUseUserContext.mockReturnValue({
      state: {
        permissions: {
          scopes: ['order:create']
        }
      }
    })
    const { container } = renderWithProviders(
      <PermissionsProvider>
        <TestingComponent />
      </PermissionsProvider>
    )
    expect(container).toMatchSnapshot()
  })
})
