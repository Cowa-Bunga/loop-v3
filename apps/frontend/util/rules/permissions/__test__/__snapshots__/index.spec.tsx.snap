// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PermissionsProvider /> should render the fleet component when the user has admin access 1`] = `
<div>
  <div>
    admin
  </div>
</div>
`;

exports[`<PermissionsProvider /> should render the fleet component when the user has fleet access 1`] = `
<div>
  <div>
    fleet
  </div>
</div>
`;

exports[`<PermissionsProvider /> should render the order create component when the order:create has admin access 1`] = `
<div>
  <div>
    order
  </div>
</div>
`;
